src/content/posts/multiplayer_online_game.md

```diff
@@ -0,0 +1,530 @@
+To resume, when you want to make a game, in which losing a single frame of a 60-FPS game is not a problem at all, and you use a way of communicating with the server that may make your client wait before it is allowed to send messages, you are definitely not using the good communication protocol.
+
+On the other hand, let me introduce you to the UDP protocol. This amazing communication protocol basically makes your communications look like this :
+- You send the only message you wanted to send : "Hello I am Zyno and happy to meet you !"
+
+And that's it ! So, obviously, you may lose some messages in the process, and you won't know it. You don't have TCP's errors detection and correction algorithms either. But as I said earlier, we do not really suffer from a lack of a message every 5 ms in a video game.
+
+### Using UDP sockets instead of TCP sockets :
+
+Now, let's go back to our code. Using UDP sockets in Python isn't really that big of a deal. In fact, it's almost the same !
+
+Look at this :
+- This is TCP :
+```
+sock = socket(AF_INET, SOCK_STREAM)
+```
+- And this is UDP :
+```
+sock = socket(AF_INET, SOCK_DGRAM)
+```
+`SOCK_STREAM` means TCP, and `SOCK_DGRAM` UDP and voilà !
+
+Ok, it is not **that** simple. The way you previously used this socket has changed a bit as well. Let's see which are the affected functions :
+
+Previously, we were using `server_sock.listen(BACKLOG)` to make a previously bound socket listen to connections, and then `sock, addr = server_sock.accept()` to make it accept connections. On the client side, we used `client_sock.connect((SERVER_IP, SERVER_PORT))` in order to connect our client socket to the listening socket. The server then generated a new socket - named sock here - and the client was now communicating with the server through its dedicated and newly created socket, using the `sock.sendall()` and `sock.recv()` functions.
+
+With UDP sockets, these functions have changed a little :
+
+There is no `sock.listen()`, `sock.accept()` nor `sock.connect()` anymore. However, the `sock.bind((IP, PORT))` function still exists and shall be used to make a socket work as a server which listen at a given IP and PORT.
+
+However, everything else now works with only two functions :
+```
+sock.sendto(bytes(message_to_send, "utf-8"), (SERVER_IP, SERVER_PORT))
+```
+And :
+```
+data, addr = sock.recvfrom(MESSAGES_LENGTH)
+
+message_received = str(data.strip(), "utf-8") # Converting the received bytes into an str
+```
+Where `message_to_send`, which is here a string, is the data to send. It is firstly converted into bytes with the utf-8 encoding. You can also send any kind of data as long as you send it using bytes. The address you send the data to is given by the second parameter : `(SERVER_IP, SERVER_PORT)`.
```

Comment:
We changed encoding because we wanted to make our sockets shorter to "reduce risks of overcrowding and socket-loss".
We also tried to mix the way our messages are encoded by using str-bytes for texts and real int-bytes for our int values to keep on reducing messages' size but it was not really successful.

Reply:
(And Chinese characters are not even writable in utf-16 so it really disappointed @BlueNZ...)

Reply:
Honestly, I would never have used utf-16 in the first place, in my opinion utf-8 is almost always the best choice,
let alone pathological case where [you use mainly non-ASCII characters](https://en.wikipedia.org/wiki/UTF-8#UTF-16) 
Quoting wikipedia:
> Text encoded in UTF-8 will be smaller than the same text encoded in UTF-16 if there are more code points below U+0080 than in the range U+0800..U+FFFF. This is true for all modern European languages. It is often true even for languages like Chinese, due to the large number of spaces, newlines, digits, and HTML markup in typical files.

Reply:
So tks for the precision 😉 you can resolve this if that's ok 👍

Reply:
Yes we realized it was a terrible idea and that's why we changed it while trying to solve our ping issues.